### ISSUE TITLE
Micrograd subtraction gradient is wrong

### ISSUE THREAD

**Issue by bobby-sills** [AayushSameerShah/Neural-Net-Zero-to-Hero-with-Andrej#1]
I'm not very experienced with calculus, but my intuition along with ChatGPT tells me that this is wrong:
``` python
def __sub__(self, other):
    other = Value(other) if not isinstance(other, Value) else other
    out = Value(self.data - other.data, (self, other), '-')
    def _backward():
        self.grad += 1.0 * out.grad
        other.grad += 1.0 * out.grad
    out._backward = _backward
    return out
```
Shouldn't the gradient of other be `other.grad += -1.0 * out.grad`?

**Comment by FloweringFor** [comment]
"other.grad += 1.0 * out.grad" is right. I think you may confuse with the difference between back propagation and gradient calculation. 

In gradient calculation, the gradient of each parameter should be accumulated. As you see in the code, "other.grad += 1.0 * out.grad".

And in back propagation, the value of the parameters is updated in the direction of gradient descent. Maybe something like that 
```python
parameter = Value(1)
...
loss.backward();
parameter.data += - lr * parameter.grad. # lr means learn rate
```
